### PR TITLE
Modified permission classes for CCX REST APIs

### LINKED ISSUE #193: The `permission_classes` of CCX REST API are not working properly
The `IsCourseInstructor` permission class seems to work fine on the detailed page but not on the list page

### PR DESCRIPTION
**Pull request by giocalitri**
This PR addresses a bug we found during the deployment on stage of https://github.com/edx/edx-platform/pull/11166. The problem was about the fact that the OAUTH user did not need to be an instructor of the master course to create CCX. Furthermore we did not take in account the fact that we want the OAUTH user to be a `staff` user on the master course and not an `instructor`.

This PR then introduces two new permission classes for the REST APIs: the first is a modified version of the old one  where we allow also `staff` members to have permission; the second is a new class that is needed to perform the check on the list view.

### Manual Testing

To test this PR, please follow the steps described in https://github.com/edx/edx-platform/pull/11166, but an extra step is to add the "dedicated user for the APIs" as a `staff member` of the master course (you can use the Instructor dashboard for this purpose).

Additional information in #193

### CODE REVIEW

File: openedx/core/lib/api/permissions.py [hunk 1]
```diff
@@ -64,13 +66,42 @@ def has_permission(self, request, view):
...
+
+
+class IsMasterCourseStaffInstructor(permissions.BasePermission):
+    """
+    Permission to check that user is instructor or staff of the master course.
+    """
+    def has_permission(self, request, view):
+        """
+        This method is assuming that a `master_course_id` parameter
+        is available in the request either as a GET parameter or it is in.
```
Comment (pdpinch) [inline openedx/core/lib/api/permissions.py]: `it is in` what? 

`it is in the request.` ?

Comment (giocalitri) [inline openedx/core/lib/api/permissions.py]: will fix

File: openedx/core/lib/api/permissions.py [hunk 2]
```diff
@@ -64,13 +66,42 @@ def has_permission(self, request, view):
...
+        The reason is because this permission class is going
+        to check if the user making the request is an instructor
+        for the specified course.
+        """
+        master_course_id = (request.GET.get('master_course_id')
+                            or request.POST.get('master_course_id')
+                            or request.data.get('master_course_id'))
```
Comment (noisecapella) [inline openedx/core/lib/api/permissions.py]: I believe `request.data` has the content from a GET or POST already

Comment (giocalitri) [inline openedx/core/lib/api/permissions.py]: only in the django rest framework environment, otherwise `request.data` does not exist

Comment (noisecapella) [inline openedx/core/lib/api/permissions.py]: Actually I'm not sure `GET` is included: http://www.django-rest-framework.org/api-guide/requests/

Comment (giocalitri) [inline openedx/core/lib/api/permissions.py]: you are right: `request.data` includes POST, PUT, PATCH but not GET

**Comment by noisecapella** [comment]
Code and tests look good, checking functionality now.

File: openedx/core/lib/api/tests/test_permissions.py [hunk 3]
```diff
@@ -16,35 +21,68 @@ def __init__(self, user=None, course_id=None):
...
+    def test_any_user_has_no_access(self):
+        self.assertFalse(self.permission.has_object_permission(self.request, None, self.obj))
+
+    def test_anonymous_has_no_access(self):
+        self.request.user = AnonymousUser()
+        self.assertFalse(self.permission.has_object_permission(self.request, None, self.obj))
+
+
+class IsMasterCourseStaffInstructorTests(TestCase):
+    """ Test for IsCourseStaffInstructor permission class. """
```
Comment (noisecapella) [inline openedx/core/lib/api/tests/test_permissions.py]: Can you update the docstring here?

Comment (giocalitri) [inline openedx/core/lib/api/tests/test_permissions.py]: sure

**Comment by giocalitri** [comment]
@noisecapella I think this is the right implementation: I added a new commit so you can see only the differences with whatever you have already reviewed

**Comment by noisecapella** [comment]
Code looks good, testing functionality now

**Comment by noisecapella** [comment]
Functionality works fine :+1:

**Comment by giocalitri** [comment]
I squashed the commits and I will open a PR against EDX shortly

**Comment by giocalitri** [comment]
Closed after opening https://github.com/edx/edx-platform/pull/11444

**Pull request closed without merging**